### ISSUE TITLE
if down the mainflow, all connect （subflow）will be shut down?

### ISSUE THREAD

**Issue by leidemon** [multipath-tcp/mptcp#383]
hi, mptcp team:
I use ubuntu 18.04.3 and mptcp 0.95 to test, and I found that if I down the mainflow, all connect will be shut down,is it normal ？How can I setup that it can switch to subflow,or it can't?
the mptcp config of client:
mptcp_path_manager     fullmesh
mptcp_scheduler             default

**Comment by leidemon** [comment]
In rfc6824, it said “ The exchange of keys in the MP_CAPABLE handshake provides material
   that can be used to authenticate the endpoints when new subflows will
   be set up”， so subflow is associated with mainflow, it is normal?

**Comment by matttbe** [comment]
Hi @leidemon 

If the first subflow of a connection is disconnected the transfer can continue on other subflows if any.

> if I down the mainflow, all connect will be shut down,is it normal?

How did you do that?

**Comment by leidemon** [comment]
hi,@matttbe:
I use the cmd "ifconfig ens3x down " to shut down the mainflow or subflow,and it behaves differently.

**Comment by leidemon** [comment]
> 嗨@leidemon
> 
> 如果连接的第一个子流断开连接，则传输可以继续进行其他子流（如果有）。
> @matttbe  So it only can switch to another subflow?  I found some cmd in the internet:
ip link set dev eth0 multipath backup
But I haven't tried.

**Comment by matttbe** [comment]
Are all subflows going through `ens3x`?

> I found some cmd in the internet: ip link set dev eth0 multipath backup

This would mark all subflows as backup. The default policy for the backup is: you only use these subflows if no other non-backup ones are available.

I guess that's not what you need.

You can also kill a subflow, like any TCP connections, using `iptables` if you need to do some experiments. Or you put an intermediate router, only used for some subflows, down.

**Comment by leidemon** [comment]
> 所有的子流都通过`ens3x`?
no，my env like this:(mptcp_fullmesh->num_subflows=1)
ens33  for mainflows 
ens38 for subflows

> > 我在因特网上找到了一些cmd：ip链接集dev th 0多路径备份。
> 
> 这将将所有子流标记为备份。备份的默认策略是：只有在没有其他非备份子流可用时才使用这些子流。
> 
> 我想这不是你需要的。
> 
> 您还可以使用以下方法杀死子流，就像任何tcp连接一样。`iptables`如果你需要做些实验的话。或者你把一个中间路由器，只用于一些子流，向下。
thank you ,I will try the iptables to kill the subflows;
rfc6824 2.2  
"The exchange of keys in the MP_CAPABLE handshake provides material
   that can be used to authenticate the endpoints when new subflows will
   be set up. "
rfc6824   3.3
“The subflow handshake mechanism (Section 3.2) will ensure that new
   subflows only join the correct connection”
it means that the subflows build on mainflows? what I care is I shut down the mainflow,the subflow will be closed too,and it does not switch to subflows?

**Comment by leidemon** [comment]
>Are all subflows going through ens3x?
no，my env like this:(mptcp_fullmesh->num_subflows=1)
ens33  for mainflows 
ens38 for subflows

>You can also kill a subflow, like any TCP connections, using iptables if you need to do some >experiments. Or you put an intermediate router, only used for some subflows, down.

thank you ,I will try the iptables to kill the subflows;
rfc6824 2.2
"The exchange of keys in the MP_CAPABLE handshake provides material
that can be used to authenticate the endpoints when new subflows will
be set up. "
rfc6824 3.3
“The subflow handshake mechanism (Section 3.2) will ensure that new
subflows only join the correct connection”
it means that the subflows build on mainflows? what I care is I shut down the mainflow,the subflow will be closed too,and it does not switch to subflows?

**Comment by matttbe** [comment]
if you have 1 MPTCP with 2 subflows, 1 per network interface, if you shutdown one interface, any of them, all the traffic should be handled by the remaining subflow.

If it is not the case, you will have to look at traces (e.g. tcpdump on each interface) to understand what happened: how the subflow have been stopped (RST? FIN?) and how the MPTCP connection is stopped (MP FastClose? MP DataFin?)

What you are trying to do is similar to: https://www.youtube.com/watch?v=VWN0ctPi5cw

**Comment by leidemon** [comment]
> if you have 1 MPTCP with 2 subflows, 1 per network interface, if you shutdown one interface, any of them, all the traffic should be handled by the remaining subflow.
> 
> If it is not the case, you will have to look at traces (e.g. tcpdump on each interface) to understand what happened: how the subflow have been stopped (RST? FIN?) and how the MPTCP connection is stopped (MP FastClose? MP DataFin?)
> 
> What you are trying to do is similar to: https://www.youtube.com/watch?v=VWN0ctPi5cw

Thank you , tomorrow I will check about it ,by the way  the config of mptcp  need some changes or just by default?

**Comment by matttbe** [comment]
you still need to configure the routing properly to use 2 different interfaces: http://multipath-tcp.org/pmwiki.php/Users/ConfigureRouting

**Comment by leidemon** [comment]
> 您仍然需要正确地配置路由以使用两个不同的接口：http://multipath-tcp.org/pmwiki.php/Users/ConfigureRouting
about this，I think it is not the problem，
I use the ss-redir and router rule to handle the packets to ss-server.then to connect to internet；
but I will check more about it.

**Comment by leidemon** [comment]
> if you have 1 MPTCP with 2 subflows, 1 per network interface, if you shutdown one interface, any of them, all the traffic should be handled by the remaining subflow.
> 
> If it is not the case, you will have to look at traces (e.g. tcpdump on each interface) to understand what happened: how the subflow have been stopped (RST? FIN?) and how the MPTCP connection is stopped (MP FastClose? MP DataFin?)
> 
> What you are trying to do is similar to: https://www.youtube.com/watch?v=VWN0ctPi5cw

How long is the waiting time for subflow switching? Can it set by config?
I take turns to disconnect one of the two subflows. Sometime it  can continue to download（I use 'nload' to check the netstat）, sometimes it don't.

**Comment by leidemon** [comment]
> 如果您有1个具有2个子流的MPTCP，每个网络接口1个，则如果关闭一个接口（其中的任何一个），则所有流量应由其余子流处理。
> 
> 如果不是这种情况，则必须查看跟踪（例如每个接口上的tcpdump）以了解发生了什么：子流如何停止（RST？FIN？）以及MPTCP连接如何停止（MP FastClose？MP）。 DataFin？）
> 
> 您要尝试执行的操作类似于：[https](https://www.youtube.com/watch?v=VWN0ctPi5cw) : [//www.youtube.com/watch?v=VWN0ctPi5cw](https://www.youtube.com/watch?v=VWN0ctPi5cw)

If I use the ifconfig to shut down the subflow, it will send the "RST" packet.
I see the youtube,it seems mptcp_path_manager  is default mode, only one dev works in the same time,
I want to test the fullmesh that it create a full-mesh of subflows among all available subflows.
and except it can switch to another one once working dev is disable.
but on my test it may have some other influences,sometimes it can continue to download, and sometimes it can not.I will test more.

**Comment by cpaasch** [comment]
What do you have in the sysctl net.ipv4.mptcp_path_manager ? Is it ```fullmesh```?

We need more logs (pcap, system-config, ip link, ip route,...)

**Comment by openaspace** [comment]
Hello.
I'm also interested to this problem that I also have. 
I start a TCP streaming connection that replicate on two LTE router, if the wan1 go down it remain working on the second, but if wan1 turn on, the connection it's not replicated newly on the wan 1.

Each wan have a static private IP to the LTE router, I need also to set static routing?

**Comment by cpaasch** [comment]
Yes, you need to make sure that the routing is configured correctly. See https://multipath-tcp.org/pmwiki.php/Users/ConfigureRouting.

**Comment by openaspace** [comment]
Thank you. I'm using OpenMPTCP router and appear to be that routing is correctly saved.
I use Cubic with Redundant , over two ethernet WAN... but I can't get to obtain wan restoration redundancy when it go offline..

![OpenMPTCProuter-Routes-LuCI](https://user-images.githubusercontent.com/13104356/95580416-96896d80-0a37-11eb-993c-8633def90770.png)

![OpenMPTCProuter-MPTCP-LuCI](https://user-images.githubusercontent.com/13104356/95580336-73f75480-0a37-11eb-97b5-8b34da320e95.png)
